gcp/sandboxes/heka/input/gcp_pubsub.lua

```diff
@@ -65,46 +68,91 @@ local err_msg = {
+
+        if cnt > 0 then
+            for i=1, cnt do
+                local msg = sdu.copy_message(default_headers, false)
+                local attrs = msgs[i][2]
+                if attrs then
+                    if attrs.heka_message then
+                        attrs.heka_message = nil
+                        msg.Uuid = attrs.Uuid
+                        attrs.Uuid = nil
+                        msg.Timestamp = int:match(attrs.Timestamp)
+                        attrs.Timestamp = nil
+                        if attrs.Hostname then
+                            msg.Hostname = attrs.Hostname
+                            attrs.Hostname = nil
+                        end
+                        if attrs.Type then
+                            msg.Type = attrs.Type
+                            attrs.Type = nil
+                        end
+                        if attrs.Logger then
+                            msg.Logger = attrs.Logger
+                            attrs.Logger = nil
+                        end
+                        if attrs.EnvVersion then
+                            msg.EnvVersion = attrs.EnvVersion
+                            attrs.EnvVersion = nil
+                        end
+                        if attrs.Severity then
+                            msg.Severity = int:match(attrs.Severity)
+                            attrs.Severity = nil
+                        end
+                        if attrs.Pid then
+                            msg.Pid = int:match(attrs.Pid)
+                            attrs.Pid = nil
+                        end
+                        local t = {}
+                        for k,v in pairs(attrs) do
+                            local k, kt = ktype:match(k)
+                            v = kt[1]:match(v)
+                            if v then
+                                t[k] = {value = v, value_type = kt[2]}
+                            end
+                        end
+                        attrs = t
                     end
+                    sdu.add_fields(msg, attrs)
                 end
-            end
-        end
-        return 0
-    end
-else
-    function process_message()
-        while is_running() do
-            local ok, msgs, cnt = pcall(subscriber.pull_sync, subscriber, batch_size)
-            if not ok then return -1, msgs end
-
-            if cnt > 0 then
-                for i=1, cnt do
-                    local ok, err = pcall(decode, msgs[i], default_headers)
-                    if not ok or err then
-                        err_msg.Payload = err
-                        err_msg.Fields.data = data
-                        pcall(inject_message, err_msg)
-                    end
+
+                local ok, err = pcall(decode, msgs[i][1], msg, true)
+                if not ok or err then
+                    err_msg.Payload = err
+                    err_msg.Fields.data = data
+                    pcall(inject_message, err_msg)
```

Comment:
does any failure here need to be handled?

Reply:
no, if we fail injecting the failure, we simply drop it